File: py/101.isSymmetric.py
# Definition for a binary tree node.
from typing import *
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
"""
review 2024.2.3
tips:
逻辑是leftChild->left == rightChild->right
     leftChild->right == rightChild->left
"""
class Solution:
    def isSymmetric(self, root: Optional[TreeNode]) -> bool:
        if not root:
            return root
        return self.traversal(root.left,root.right)
    def traversal(self,l:TreeNode,r:TreeNode):
        if not l and not r:
            return True
        if not l and r:
            return False
        if l and not r:
            return False
        if l.val != r.val:
            return False
        return self.traversal(l.left,r.right) and self.traversal(l.right,r.left)

File: java/101.isSymmetric.java

//  Definition for a binary tree node.
public class TreeNode {
      int val;
      TreeNode left;
      TreeNode right;
      TreeNode() {}
      TreeNode(int val) { this.val = val; }
      TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    public boolean isSymmetric(TreeNode root) {
        return recursion(root,root);
    }
    public boolean recursion(TreeNode l,TreeNode r) {
        if (l == null && r != null) 
            return false;
        if (l != null && r == null)
            return false;
        if (l == null && r == null)
            return true;
        if (l.val != r.val)
            return false;
        return recursion(l.left,r.right) && recursion(l.right,r.left);
    }
}